src/expr/src/scalar/func.rs

```diff
@@ -1750,6 +1755,49 @@ fn date_part_interval_inner(
     }
 }
 
+fn date_part_time<'a, T>(a: Datum<'a>, time: T) -> Result<Datum<'a>, EvalError>
+where
+    T: TimeLike
+{
+    let units = a.unwrap_str();
+    match units.parse() {
+        Ok(units) => date_part_time_inner(units, time),
+        Err(_) => Err(EvalError::UnknownUnits(units.to_owned())),
+    }
+}
+
+fn date_part_time_inner<'a, T>(units: DateTimeUnits, time: T) -> Result<Datum<'a>, EvalError>
+where
+    T: TimeLike
+{
+    match units {
+        DateTimeUnits::Hour => Ok(time.extract_hour().into()),
+        DateTimeUnits::Minute => Ok(time.extract_minute().into()),
+        DateTimeUnits::Second => Ok(time.extract_second().into()),
+        DateTimeUnits::Milliseconds => Ok(time.extract_millisecond().into()),
+        DateTimeUnits::Microseconds => Ok(time.extract_microsecond().into()),
+        DateTimeUnits::Epoch
+        | DateTimeUnits::Year
+        | DateTimeUnits::Week
+        | DateTimeUnits::Day
+        | DateTimeUnits::Millennium
+        | DateTimeUnits::Century
+        | DateTimeUnits::Decade
+        | DateTimeUnits::Quarter
+        | DateTimeUnits::Month
+        | DateTimeUnits::DayOfWeek
+        | DateTimeUnits::DayOfYear
+        | DateTimeUnits::IsoDayOfWeek
+        | DateTimeUnits::IsoDayOfYear => Err(EvalError::UnknownUnits(format!("{}", units))),
```

Comment:
Let me know if `UnknownUnits` is an appropriate error here. I was also thinking maybe `InvalidParameterValue`, but I wasn't sure.

Reply:
Whatever matches PostgreSQL is the way to go, even if that means introducing a new error variant!

Reply:
Ok, so I looked around a bit and I think that `UnkownUnits` is the correct error. In materialize `UnknownUnits` is used for unit errors in the `date_part`, `date_trunc`, and `date_bin` functions.

In Materialize this error message ends up looking like: `ERROR: unkown units '{unit}'` (ex: 
```
materialize=> SELECT date_trunc('hoourr', TIMESTAMP '2001-02-16 20:38:40');
ERROR:  unknown units 'hoourr'
```
)

In postgres this error message ends up looking like `ERROR:  {type} units "{unit}" not recognized` (ex:
```
postgres=# SELECT date_trunc('hoourr', TIMESTAMP '2001-02-16 20:38:40');
ERROR:  timestamp units "hoourr" not recognized
```
)

If it's a small change I'll try and make the error message in Materialize include the type so that it matches postgres. Otherwise it might be best fixed in a followup issue/PR.

As a side note sometimes Postgres includes quotes (") around the type, ex: `ERROR:  timestamp units "hoourr" not recognized` vs `ERROR:  "time" units "hoourr" not recognized`. I'm not sure why they have this difference and I've only ever noticed the quotes when the type is `time`. Doesn't seem important enough to replicate exactly, though let me know if you disagree.

Reply:
Huh! Good eye. Yeah, as long as you're using the error type that *should* match PostgreSQL, that's good enough for this PR. Matching error messages exactly is pretty far down on the list of important PostgreSQL compat stuff.

The quoting of `"time"` and `"time with time zone"` appears to be just an inconsistency in the PostgreSQL error messages: https://github.com/postgres/postgres/blob/95ab1e0a9db321dd796344d526457016eada027f/src/backend/utils/adt/date.c#L2208

Reply:
Sounds good, thanks for the feedback. In that case I'll resolve this conversation.